jobs/audit_jobs.py

```diff
@@ -43,70 +45,58 @@ def run(self):
+                # requirement.
+                len(KIND_BY_INDEX), KIND_BY_INDEX)
         )
 
-        audits_by_kind = audit_decorators.AuditsExisting.get_audits_by_kind()
-        kind_and_audit_class_pairs = (
-            self.pipeline
-            | 'Compute the predefined (kind, audit class) pairs' >> beam.Create(
-                (kind, audit_class)
-                for kind, audit_classes in audits_by_kind.items()
-                for audit_class in audit_classes)
-        )
+        audit_error_pcolls = list(itertools.chain.from_iterable(
+            (self._run_do_fn(do_fn, model_kind, models) for do_fn in do_fns)
+            for do_fns, model_kind, models in python_utils.ZIP(
+                AUDIT_DO_FNS_BY_INDEX, KIND_BY_INDEX, models_of_kind_by_index)
+            if do_fns and models))
 
-        existing_model_errors = (
-            (kind_and_existing_model_pairs, kind_and_audit_class_pairs)
-            | 'Join models and audit classes by kind' >> beam.CoGroupByKey()
-            | 'Filter empty joins' >> beam.Filter(self._is_non_empty_join)
-            | 'Audit the models' >> beam.FlatMap(self._audit_models)
-        )
-
-        deleted_model_errors = (
+        audit_error_pcolls.append(
             deleted_models
-            | 'Run ValidateDeletedModel on deleted models' >> beam.ParDo(
-                base_model_audits.ValidateDeletedModel())
-        )
-
-        return (
-            (existing_model_errors, deleted_model_errors)
-            | 'Combine errors' >> beam.Flatten()
+            | 'Run ValidateDeletedModel on deleted models' >> (
+                beam.ParDo(base_model_audits.ValidateDeletedModel()))
         )
 
-    def _audit_models(self, join_result):
-        """Runs each of the audit DoFns on each of the models.
-
-        Args:
-            join_result: tuple(str, tuple(PCollection, PCollection)). The output
-                from CoGroupByKey.
-
-        Returns:
-            PCollection. An aggregate PCollection of errors from the audits.
-        """
-        kind, (models, audit_classes) = join_result
-        audit_error_pcolls = (
-            models | 'Run %s on %ss' % (cls.__name__, kind) >> beam.ParDo(cls())
-            for cls in audit_classes
-        )
-        return audit_error_pcolls | beam.Flatten()
+        return audit_error_pcolls | 'Combine all audit errors' >> beam.Flatten()
 
-    def _is_non_empty_join(self, join_result):
-        """Returns whether the joined pair of (models, audit_classes) are empty.
+    def _run_do_fn(self, do_fn, model_kind, models):
+        """Runs a DoFn over the specified kind of models.
 
         Args:
-            join_result: tuple(str, tuple(PCollection, PCollection)). The output
-                from CoGroupByKey.
+            do_fn: DoFn. The type of DoFn to run.
+            model_kind: str. The kind of model the DoFn is processing.
+            models: PCollection. The models for the DoFn to process.
```

Comment:
What's the difference between a PCollection-type `models` and a regular list of datastore models? It might perhaps be worth saying a bit about that in the docstring -- e.g. "The collection of models for ..." since just "The models" typically implies list(models) / set(models).